Add ability to optionally split into multiple ERDs

craineum opened the pull request:
Really love this tool! It actually caught an issue with a project I am working on because it failed when trying to build the file with a reference that was invalid because of something that had gotten removed, but not everything was cleaned up.

Unfortunately, that is kinda all that I am able to do with this gem. Our generated mermaid file is too large. So I thought it would be cool to add the ability to output multiple files based on the different entities and some of their close relations. This would keep the files small and manageable. I also added a navigation element to switch between the various entities.

Right now this is just a stab at making this work. There is still a little more to do, but I wanted to get your feedback. This is a pretty aggressive change, so completely understand if it is too much.

Let me know your thoughts.

Details of changes so far:
- Add configuration option to enable split output
- Outputs one file per model with immediate relations
- Still need to implement relationship depth configuration
- Added data to test suite to validate

craineum committed:
Add ability to optionally split into multiple ERDs

- Add configuration option to enable
- Outputs one file per model with immediate relations
- Still need to implement depth configuration to show more relations
- Added data to test suite to validate

humzahkiani commented:
Thanks for the support! I only perused your PR, but it seems like you are attempting to add the ability to generate ERDs localized around a given model, with a configured search depth similar to a Breadth-First Search. Is that correct?

craineum commented:
> Thanks for the support! I only perused your PR, but it seems like you are attempting to add the ability to generate ERDs localized around a given model, with a configured search depth similar to a Breadth-First Search. Is that correct?

Yup, that is exactly what I was going for.

craineum committed:
Add depth option to split output

humzahkiani commented:
Cool, I like this idea. ERDs are a visual tool for people to look at. As the number of models increases and complexity grows, you lose value in what can be understood from the tool.  

Feel free to continue working on it. I may take some issue with the specific implementation details, but I'll be sure to let you know in the PR. I'm a bit busy at the moment, but should be able to take a closer look within the next couple weeks.

craineum commented:
Sounds good, completely understand giving input and I am completely open to anything. Once I am happy with what I have I will promote it from a draft PR.

humzahkiani commented on lib/rails-mermaid_erd_markdown.rb:
```diff
@@ -14,59 +14,51 @@ module MermaidErdMarkdown
+      lines << "## Models"
+      lines << ""
+      files.each do |model, path|
+        lines << "- [#{model}](#{path})"
```
The index ERD file is also in the top level of the path, which means that this path won't work. 
Prepending a `../` should do it.

craineum replied:
This should be okay since I am appending an `_index` to the filename.

craineum replied:
Oh... it is outputting `docs/thing` and that doesn't link from `docs/index`, got it!

humzahkiani commented on lib/rails-mermaid_erd_markdown.rb:
```diff
@@ -14,59 +14,51 @@ module MermaidErdMarkdown
-      if Digest::MD5.hexdigest(current_erd) == Digest::MD5.hexdigest(new_erd)
-        logger.info("ERD already exists and is up to date. Skipping...")
-        return
+    def index_markdown(files)
```
Adding this is a nice touch

humzahkiani commented:
Played around with the PR for a bit. Seems to work for a basic case with a few models.
I'll do more edge case testing when you feel ready to promote the PR. Otherwise, looks great.

craineum committed:
Separate concerns & update version

- Create `Markdown` class to handle all markdown generation
- Move functionality to `Document` class to create instance

craineum commented:
This last commit separated the concerns a bunch and added some new classes to support that. Going to test this on some larger projects and see how it handles it. Once I am good with that I will mark it ready. If you have concerns about how I structured things or any of the changes please let me know. I am totally open to feedback!

craineum committed:
Remove relation without model from split output

- Fix logger

craineum commented:
Okay, I am pretty happy with the results of my testing. There is some weird stuff that happens in one particular codebase that uses lots of model inheritance, but nothing that breaks it. Moving this to ready for your review.

humzahkiani commented on lib/rails-mermaid_erd_markdown/markdown.rb:
```diff
@@ -0,0 +1,105 @@
+    end
+
+    def generate(&block)
+      block.arity == 1 ? block[self] : instance_eval(&block)
```
A bit confused about what's happening here

craineum replied:
First it checks if the passed in block is expecting an attribute, then yields to the block in one of two ways.
- With no expected attribute it will process the block as if it is running within the `Markdown` class able to make calls like directly to methods within via `insatnace_eval(&block)`, but would not have access to the calling classes methods like:
```ruby
def body
  "Description"
end

def new_markdown
  MermaidErdMarkdown::Markdown.generate do
    add header("Title") #This would work as it is in scope in the `Markdown` class.
    add body #This would fail as body isn't in scope here.
  end
end
```
- With an attribute passed then I have access to both since `self` is the thing the attribute is being set to. Then you reference the `Markdown` class with the attribute:
```ruby
def body
  "Description"
end

def new_markdown
  MermaidErdMarkdown::Markdown.generate do |doc|
    doc.add doc.header("Title")
    doc.add body #This now works.
  end
end
```

I started with the former first, then changed things to the latter as I needed. Looking back I probably could convert the `mermaid_markdown`'s block to not used the attribute, but that might be more confusing. I could clean this up so it expects an arity of 1 and get rid of the other.

humzahkiani replied:
How does `block[self]` have access to the surrounding class context? Doesn't `self` here also refer to the `Markdown` instance, since it is called within `Markdown`?

craineum replied:
It is passing self back to the calling method as the param used there.  So in the example above the `Markdown` class becomes the `doc` param. This encapsulation results in the rest of the block running in the calling instance, thus having access to the rest of the calling instance. And `doc` having access to the `Markdown` instance.

humzahkiani commented:
Testing looks good for me as well. I noticed some issues with ActiveRecord classes not being loaded correctly in certain rails apps but that is unrelated to your PR. 

For the structuring of your classes, I'm curious about why you organized it as you did compared to something like this:

```
class BaseDocument
      abstract
      def add_header
      def add_row
      
class MarkdownDocument < BaseDocument
      def add_header
           # Override with markdown specific logic
      def add_row

class Performer
      def self.generate
             new.generate
      def generate
             create_comprehensive_markdown_erd
             @erd = ERD.new
          
     def  create_comprehensive_markdown_erd
             doc = MarkdownDocument.new
             doc.add_row(process(@erd.models))
             
class ERD
	def models
		@models ||= ERDMarkdown.model_data  
```

It feels a bit strange to generate (markdown) documents within the document Class. I'm relatively new to ruby/rails so always curious about best practices and conventions.

craineum commented:
@humzahkiani I think that comes down to preference. I separated concerns, but didn't break it down to that level. I put everything that dealt with generating ERD markdown into one file, because that is all that is currently needed. I have no issues implementing what you are proposing if that would increase maintainability/readability.

humzahkiani commented:
@craineum I'm realizing that the design I proposed would make the gem more extensible for different file types (non-Markdown) and data concerns (non-ERD). 

Considering the gem is called `rails-mermaid_erd_markdown`, these changes would be out of scope unless I wished to redefine the gem's function. For now I'd prefer it to be well-defined and single-purpose.

For readability, I think we might benefit from having the design be a bit more intuitive. 
To me, in your code, the `Markdown` class is more of the domain for generating markdown documents, and the `Document` class is more of a orchestrator/performer for creating/updating/configuring the different ERD Markdown documents in the codebase. 

What do you think about changing the class names to fit this, and extracting out the ERD data into it's own class? 
Not the erd updating logic but the interface for the data itself, such as `models`, `related_models`, `related_model_names` etc.

craineum commented:
@humzahkiani Yup, I like this idea. I will give it a go.

craineum committed:
Update Gemfile.lock

craineum committed:
Fix test fixture

craineum committed:
Change generate markdown for mermaid doc

craineum committed:
Separate source data handling from `Document`.

- Move test `util` directory to `example_output` for clarity.

craineum committed:
Rename classes for better clarity

craineum commented:
@humzahkiani Okay, changes complete! Let me know if there are any changes you want made.

humzahkiani commented:
Looks great! The only thought I have is it would be amazing if we could get even more granular about which models generate ERDs, and configure specific relationship depths for each. 

But I am happy to do that myself in a a subsequent version.
Approved. Appreciate the work you've done here.
